src/rust/graph-generator-lib/src/aws.rs

```diff
@@ -0,0 +1,221 @@
+use crate::serialization::SubgraphSerializer;
+use aws_lambda_events::event::sqs::SqsEvent;
+use graph_descriptions::graph_description::*;
+use grapl_config as config;
+use lambda_runtime::error::HandlerError;
+use lambda_runtime::Context;
+use log::*;
+use rusoto_core::Region;
+use rusoto_s3::S3Client;
+use rusoto_sqs::SqsClient;
+use rusoto_sts::{StsAssumeRoleSessionCredentialsProvider, StsClient};
+use sqs_lambda::event_decoder::PayloadDecoder;
+use sqs_lambda::event_handler::EventHandler;
+use std::collections::HashSet;
+use std::str::FromStr;
+use std::sync::mpsc::SyncSender;
+use std::time::Duration;
+
+/// Runs the graph generator on AWS
+///
+/// This function will assume the generator is running on an AWS lambda with services such as
+/// S3 and SQS immediately available and accessible.
+pub(crate) fn run_graph_generator_aws<
+    IE: Send + Sync + Clone + 'static,
+    EH: EventHandler<InputEvent = IE, OutputEvent = Graph, Error = sqs_lambda::error::Error>
+        + Send
+        + Sync
+        + Clone
+        + 'static,
+    ED: PayloadDecoder<IE> + Send + Sync + Clone + 'static,
+>(
+    generator: EH,
+    event_decoder: ED,
+) {
+    lambda_runtime::lambda!(|event, context| {
+        lambda_handler(event, context, generator.clone(), event_decoder.clone())
+    })
+}
+
+fn lambda_handler<
+    IE: Send + Sync + Clone + 'static,
+    EH: EventHandler<InputEvent = IE, OutputEvent = Graph, Error = sqs_lambda::error::Error>
+        + Send
+        + Sync
+        + Clone
+        + 'static,
+    ED: PayloadDecoder<IE> + Send + Sync + Clone + 'static,
+>(
+    event: SqsEvent,
+    ctx: Context,
+    generator: EH,
+    event_decoder: ED,
+) -> Result<(), HandlerError> {
+    info!("Handling event");
+
+    // tracks the SQS message ids in this batch of events
+    let mut initial_events: HashSet<String> = event
+        .records
+        .iter()
+        .map(|event| event.message_id.clone().unwrap())
+        .collect();
+
+    info!("Initial Events {:?}", initial_events);
+
+    let (tx, rx) = std::sync::mpsc::sync_channel(10);
+
+    let t = std::thread::spawn(move || {
+        // lambda_runtime uses tokio::runtime::Runtime internally so we need to use
+        // tokio_compat::run_std if we want to invoke a new async task
+        tokio_compat::run_std(run_async_generator_handler(
```

Comment:
This used to be a closure but it was very hard to parse so I moved that out into its own function.